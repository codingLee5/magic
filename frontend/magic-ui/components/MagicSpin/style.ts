import { createStyles } from "antd-style"
import { useMemo } from "react"

const sizeMap = {
	small: 30,
	default: 50,
	large: 70,
}

export const useStyles = createStyles(({ css, prefixCls }, { size }: { size: number }) => {
	return {
		wrapper: css`
			width: 100%;
			// .${prefixCls}-spin-blur.${prefixCls}-spin-blur {
			// 	opacity: 0;
			// }
			display: flex;
			justify-content: center;
			align-items: center;

			.${prefixCls}-spin-container {
				width: 100%;
				height: 100%;
			}
		`,
		icon: css`
			--${prefixCls}-spin-content-height: unset !important;

			.${prefixCls}-spin-dot {
				width: ${size}px;
				height: ${size}px;
				--magic-spin-dot-size: ${size}px;
			}

			&.${prefixCls}-spin-show-text > .${prefixCls}-spin-dot {
				transform: translateY(-50%);
			}
		`,
	}
})

export function useMagicSpinProps(size?: "small" | "default" | "large") {
	const { styles } = useStyles({ size: sizeMap[size ?? "default"] })

	return useMemo(
		() => ({
			rootClassName: styles.icon,
			wrapperClassName: styles.wrapper,
		}),
		[styles.icon, styles.wrapper],
	)
}
