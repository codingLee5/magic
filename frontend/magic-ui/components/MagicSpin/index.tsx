import type { SpinProps } from "antd"
import { Spin } from "antd"
import { memo } from "react"
import { useMagicSpinProps } from "./style"
import { cx } from "antd-style"

export type MagicSpinProps = SpinProps

const MagicSpin = memo(function MagicSpin({ children, size, className, ...props }: MagicSpinProps) {
	const magicSpinProps = useMagicSpinProps(size)

	return (
		<Spin
			{...magicSpinProps}
			wrapperClassName={cx(magicSpinProps.wrapperClassName, className)}
			{...props}
		>
			{children}
		</Spin>
	)
})

export default MagicSpin
