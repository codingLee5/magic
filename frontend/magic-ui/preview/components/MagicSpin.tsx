import React from "react"
import { Space, Card } from "antd"
import MagicSpin from "../../components/MagicSpin"
import ComponentDemo from "./Container"

const SpinDemo: React.FC = () => {
	return (
		<div>
			<ComponentDemo title="基础加载" description="最基本的加载组件" code="<MagicSpin />">
				<Space size={100}>
					<MagicSpin />
					<MagicSpin size="large" />
					<MagicSpin size="small" />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="包裹内容的加载"
				description="将加载状态包裹在内容周围"
				code="<MagicSpin><div>内容</div></MagicSpin>"
			>
				<MagicSpin tip="加载中...">
					<Card title="内容卡片" style={{ width: 300 }}>
						<p>这是被加载组件包裹的内容</p>
						<p>当loading为true时会显示加载动画</p>
					</Card>
				</MagicSpin>
			</ComponentDemo>

			<ComponentDemo
				title="不同尺寸"
				description="支持不同尺寸的加载动画"
				code="size: 'small' | 'default' | 'large'"
			>
				<Space size={100}>
					<MagicSpin size="small" tip="小尺寸" />
					<MagicSpin tip="默认尺寸" />
					<MagicSpin size="large" tip="大尺寸" />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="延迟显示"
				description="延迟显示加载动画，避免闪烁"
				code="delay: 500"
			>
				<Space size={100}>
					<MagicSpin delay={500} tip="延迟500ms显示" />
					<MagicSpin delay={1000} tip="延迟1秒显示" />
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default SpinDemo
