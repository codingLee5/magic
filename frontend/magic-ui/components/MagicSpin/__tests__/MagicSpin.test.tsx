import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicSpin from "../index"

describe("MagicSpin", () => {
	it("should render with children", () => {
		render(
			<MagicThemeProvider theme="light">
				<MagicSpin>
					<div>内容区域</div>
				</MagicSpin>
			</MagicThemeProvider>,
		)

		// 检查是否渲染了子元素
		expect(screen.getByText("内容区域")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础加载器快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin>
						<div>内容区域</div>
					</MagicSpin>
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("加载状态快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin spinning>
						<div>内容区域</div>
					</MagicSpin>
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带提示文字快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin tip="加载中...">
						<div>内容区域</div>
					</MagicSpin>
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("小尺寸加载器快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin size="small">
						<div>内容区域</div>
					</MagicSpin>
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("大尺寸加载器快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin size="large">
						<div>内容区域</div>
					</MagicSpin>
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("延迟加载快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin delay={500}>
						<div>内容区域</div>
					</MagicSpin>
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("独立加载器快照", () => {
			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSpin />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
