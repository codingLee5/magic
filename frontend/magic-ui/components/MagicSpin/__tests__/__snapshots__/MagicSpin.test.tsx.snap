// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicSpin > 快照测试 > 加载状态快照 1`] = `
<DocumentFragment>
  <div
    class="magic-spin-nested-loading acss-1tvuebw css-dev-only-do-not-override-yotw4n css-var-r2"
  >
    <div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="magic-spin magic-spin-spinning acss-sxthw7 css-dev-only-do-not-override-yotw4n css-var-r2"
      >
        <span
          class="magic-spin-dot-holder"
        >
          <span
            class="magic-spin-dot magic-spin-dot-spin"
          >
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
          </span>
        </span>
      </div>
    </div>
    <div
      class="magic-spin-container magic-spin-blur"
    >
      <div>
        内容区域
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicSpin > 快照测试 > 基础加载器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-spin-nested-loading acss-1tvuebw css-dev-only-do-not-override-yotw4n css-var-r1"
  >
    <div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="magic-spin magic-spin-spinning acss-sxthw7 css-dev-only-do-not-override-yotw4n css-var-r1"
      >
        <span
          class="magic-spin-dot-holder"
        >
          <span
            class="magic-spin-dot magic-spin-dot-spin"
          >
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
          </span>
        </span>
      </div>
    </div>
    <div
      class="magic-spin-container magic-spin-blur"
    >
      <div>
        内容区域
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicSpin > 快照测试 > 大尺寸加载器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-spin-nested-loading acss-1tvuebw css-dev-only-do-not-override-yotw4n css-var-r5"
  >
    <div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="magic-spin magic-spin-spinning acss-znmur6 css-dev-only-do-not-override-yotw4n css-var-r5"
      >
        <span
          class="magic-spin-dot-holder"
        >
          <span
            class="magic-spin-dot magic-spin-dot-spin"
          >
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
          </span>
        </span>
      </div>
    </div>
    <div
      class="magic-spin-container magic-spin-blur"
    >
      <div>
        内容区域
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicSpin > 快照测试 > 小尺寸加载器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-spin-nested-loading acss-1tvuebw css-dev-only-do-not-override-yotw4n css-var-r4"
  >
    <div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="magic-spin magic-spin-spinning acss-12hgqwl css-dev-only-do-not-override-yotw4n css-var-r4"
      >
        <span
          class="magic-spin-dot-holder"
        >
          <span
            class="magic-spin-dot magic-spin-dot-spin"
          >
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
          </span>
        </span>
      </div>
    </div>
    <div
      class="magic-spin-container magic-spin-blur"
    >
      <div>
        内容区域
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicSpin > 快照测试 > 带提示文字快照 1`] = `
<DocumentFragment>
  <div
    class="magic-spin-nested-loading acss-1tvuebw css-dev-only-do-not-override-yotw4n css-var-r3"
  >
    <div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="magic-spin magic-spin-spinning magic-spin-show-text acss-sxthw7 css-dev-only-do-not-override-yotw4n css-var-r3"
      >
        <span
          class="magic-spin-dot-holder"
        >
          <span
            class="magic-spin-dot magic-spin-dot-spin"
          >
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
            <i
              class="magic-spin-dot-item"
            />
          </span>
        </span>
        <div
          class="magic-spin-text"
        >
          加载中...
        </div>
      </div>
    </div>
    <div
      class="magic-spin-container magic-spin-blur"
    >
      <div>
        内容区域
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicSpin > 快照测试 > 延迟加载快照 1`] = `
<DocumentFragment>
  <div
    class="magic-spin-nested-loading acss-1tvuebw css-dev-only-do-not-override-yotw4n css-var-r6"
  >
    <div
      class="magic-spin-container"
    >
      <div>
        内容区域
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicSpin > 快照测试 > 独立加载器快照 1`] = `
<DocumentFragment>
  <div
    aria-busy="true"
    aria-live="polite"
    class="magic-spin magic-spin-spinning acss-sxthw7 css-dev-only-do-not-override-yotw4n css-var-r7"
  >
    <span
      class="magic-spin-dot-holder"
    >
      <span
        class="magic-spin-dot magic-spin-dot-spin"
      >
        <i
          class="magic-spin-dot-item"
        />
        <i
          class="magic-spin-dot-item"
        />
        <i
          class="magic-spin-dot-item"
        />
        <i
          class="magic-spin-dot-item"
        />
      </span>
    </span>
  </div>
</DocumentFragment>
`;
